CoreCLR: vm: arm: Fix gcc arm asm build error

giuliobenetti opened the pull request:
Building with gcc on Arm architecture leads to error due to missing '0x' prefix on hexadecimal values

giuliobenetti commented:
@dotnet-policy-service agree company="Benetti Engineering"

giuliobenetti commented:
> hello @giuliobenetti, do you plan on fixing the CI build breaks so we could complete this PR?

Hello @mangod9,
I spent some time trying to reproduce the build failure but I haven't made it because of lack of time.
Can someone else add the -fPIC to CFLAGS?

Thank you!

mangod9 commented:
this doesnt appear critical for 8, so we will take a look after fork.

janvorli commented:
@giuliobenetti this change is incorrect. The `1f` and similar are not hexadecimal numbers. These represents numbered local labels, the `f` means "forward" and `b` means "backward". The assembly code can have multiple labels with the same number, so the `f` and `b` are used to disambiguate them.

giuliobenetti commented:
> @giuliobenetti this change is incorrect. The `1f` and similar are not hexadecimal numbers. These represents numbered local labels, the `f` means "forward" and `b` means "backward". The assembly code can have multiple labels with the same number, so the `f` and `b` are used to disambiguate them.

Thanks a lot @janvorli, I didn’t know this. I will fix it then. Maybe that’s the reason why I experience a segfault very early after building from scratch on Buildroot.
Kind regards

giuliobenetti committed:
src/coreclr/vm/arm/asmhelpers.S: fix building with GNU AS

GNU AS doesn't support label subtraction when assigning to a register so
let's load absolute function address to R2 when building with GNU AS.

Signed-off-by: Giulio Benetti <giulio.benetti+tekvox@benettiengineering.com>

giuliobenetti committed:
src/coreclr/vm/arm/pinvokestubs.S: fix building with GNU AS

GNU AS doesn't support label subtraction when assigning to a register so
let's load absolute function address to R1/R2 when building with GNU AS.

Signed-off-by: Giulio Benetti <giulio.benetti+tekvox@benettiengineering.com>

giuliobenetti commented:
Hello @janvorli,

I've updated the 2 commits with something that is not a real solution.
I've found out the reason why you've committed the (1f+4) patches on this PR:
https://github.com/dotnet/runtime/commit/39803d4d3cbf818641437afb7402da15fa076bec
But unfortunately GAS doesn't support label subtraction on the fly when:
ldr     r2, =JIT_WriteBarrier_Loc-(1f+4)
so I've guarded by __clang__ and used absolute addresses instead of relative ones.
In the PR you talk about ARM64 RHEL showing up the problem of RX/RO, have you experienced that on ARM32 too?
My goal is to add dotnet-runtime as a Buildroot package, but there we only have LLVM very OpenCL oriented only and not available as toolchain.
So I have to go with GNU. Maybe I can try to find another workaround for this.
What do you think?

janvorli commented:
The way you've made it for GAS would produce runtime relocations, the subtraction of labels prevents that - the delta is constant and it is then added to PC. Without the subtraction, the actual address has to be computed at runtime based on where the libcoreclr.so got loaded. I wonder if using non-numeric local labels would allow the subtraction to work on GAS. I would be really surprised if that would not be possible there.
What I mean is transforming the code like
```
    ldr     r2, =JIT_WriteBarrier_Loc-(1f+4) // or R3? See targetarm.h
1:
    add     r2, pc
```
to
```
    ldr     r2, =JIT_WriteBarrier_Loc-(LOCAL_LABEL(L1)+4) // or R3? See targetarm.h
LOCAL_LABEL(L1):
    add     r2, pc
```

As for the RO/RW issue, that was purely arm64 issue due to the 64kB memory pages. That issue has nothing to do with the relocations in the asm code.

giuliobenetti commented:
> 




> The way you've made it for GAS would produce runtime relocations, the subtraction of labels prevents that - the delta is constant and it is then added to PC. Without the subtraction, the actual address has to be computed at runtime based on where the libcoreclr.so got loaded. 

Ok, so for Arm32 it’s a performance issue and I should address it in some way.

> I wonder if using non-numeric local labels would allow the subtraction to work on GAS. I would be really surprised if that would not be possible there. 

Unfortunately it is exactly like that up to binutils 2.39 I haven’t tried with 2.40 and 2.41 to be honest, but I’d prefer to find a more canonical way to do that.

> What I mean is transforming the code like
> 
> ```
>     ldr     r2, =JIT_WriteBarrier_Loc-(1f+4) // or R3? See targetarm.h
> 1:
>     add     r2, pc
> ```
> 
> to
> 
> ```
>     ldr     r2, =JIT_WriteBarrier_Loc-(LOCAL_LABEL(L1)+4) // or R3? See targetarm.h
> LOCAL_LABEL(L1):
>     add     r2, pc
> ```

This ^^^ could be a canonical way. But GAS doesn’t support labels subtraction :-/ I need a workaround for this. I’m going to try to find it.
I have a simple test case for GAS and then I check with objdump —disassemble to see the result.

Anyway I’m running with relocations on a i.MX6ULL@528Mhz and I experience more or less 20% less cpu usage compared to prebuilt image if I’m correct. Now it’s working most of all and I plan to send a patch for dotnet-runtime to Buildroot soon. This way we obtain an easy way to build taking into account all the possible tweaks(fpu, optimizations etc.). I’ve given a go on arm32+glibc but I plan to test it on aarch64 and musl too and step by step(first ones x86/x64) other available architecture in qemu because of lack of hardware.

> As for the RO/RW issue, that was purely arm64 issue due to the 64kB memory pages. That issue has nothing to do with the relocations in the asm code.

Oh ok, that sounds clear to me now.

Thanks a lot for the explanation!

giuliobenetti commented:
@janvorli last thing. Maybe we can consider these two patches ok for the moment since they fix building with gcc at the cost of loosing performance. But at least it builds fine and in the end I find it to be faster this way compared to prebuilt tarballs. Then later yes I can improve it following your explanation above.

It would be more likely accepted in Buildroot this way.

What about this?

mangod9 commented:
@giuliobenetti could you please rebase so we can run CI again. @janvorli are there more changes required here?

giuliobenetti commented:
> @giuliobenetti could you please rebase so we can run CI again. @janvorli are there more changes required here?

Just done. 

> @giuliobenetti could you please rebase so we can run CI again. @janvorli are there more changes required here?

Just pushed --force. Once it works for main branch I'm going to open PRs for net6 and net7 branches. I'm planning to send a patch to Buildroot for .NET6/7/8 so one can build it using it. I've successfully built runtime 6 and need some clean up for 7 and 8 to build successfully. I'd also need this PR:
https://github.com/dotnet/runtime/pull/90554
 merged in order to build in Buildroot, can someone take care of it?

janvorli commented:
@giuliobenetti I am sorry, I've missed the last two comments due to my vacation back in August. 
Your latest update seems like a bad merge, it is back to the incorrect replacement of `1f` by `0x1f` etc.

giuliobenetti commented:
> @giuliobenetti I am sorry, I've missed the last two comments due to my vacation back in August.
> Your latest update seems like a bad merge, it is back to the incorrect replacement of `1f` by `0x1f` etc.

Oops. In 30 minutes i update with correct patches.

giuliobenetti commented:
@janvorli it looks like some build test hung. Does the code looks correct now? Unfortunately GCC doesn't support the difference of labels addresses, so I had to go with #ifdefs. But overall if I build runtime with Buildroot the performance looks better than using the prebuilt images.

janvorli commented:
@giuliobenetti my concern w.r.t. the relocations was not primarily the performance, but rather security. See https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro for details. So I wanted to check what we get with this change.
To my surprise, your version built with GCC is still a full relro:
```
checksec --file=artifacts/bin/coreclr/linux.arm.Debug/libcoreclr.so
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      Symbols         FORTIFY Fortified       Fortifiable     FILE
Full RELRO      Canary found      NX enabled    DSO             No RPATH   No RUNPATH   352383) Symbols   No    0               22              artifacts/bin/coreclr/linux.arm.Debug/libcoreclr.so
```
So the change looks good then. Just out of my interest, I've tried to use your version with clang / lld, but the linker failed stating that there is a relocation not compatible with relro. So we'll need to keep it split.

giuliobenetti commented:
@janvorli thanks a lot for the explanation. Now it’s clear to me the reason of the shared library function call with the label trick.
Buildroot set by default full RELRO for toolchains and probably GAS or LD do the job for meet full RELRO. This is my understanding considering that I’ve never dealt with such security bound code, but it’s very interesting, so thanks again for pointing me that link! And here I have a practical example on how it could impact, also a particular example that involves assembly.
Let’s see if pipelines succeeds then :-)

Kind regards!

giuliobenetti commented:
@janvorli do you have news for this PR?

Thanks in advance

janvorli commented:
@giuliobenetti I am very sorry, this PR went out of my radar completely. I will merge it in a minute.

janvorli reviewed (approved):
LGTM, thank you!

giuliobenetti commented:
> @giuliobenetti I am very sorry, this PR went out of my radar completely. I will merge it in a minute.

No problem :-)
